Doxygen for torchbind

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35007 Doxygen for torchbind**

![image](https://user-images.githubusercontent.com/4685384/77018550-1e444900-693b-11ea-8dfa-828e764130b2.png)
![image](https://user-images.githubusercontent.com/4685384/77018556-23a19380-693b-11ea-9e71-5c07edd19ff5.png)
![image](https://user-images.githubusercontent.com/4685384/77018564-2bf9ce80-693b-11ea-897a-16972687f2a4.png)
![image](https://user-images.githubusercontent.com/4685384/77018573-31571900-693b-11ea-9852-065f7f99fd2c.png)


Differential Revision: [D20525680](https://our.internmc.facebook.com/intern/diff/D20525680)

Commit:
Doxygen for torchbind

[ghstack-poisoned]

Commit:
Update on "Doxygen for torchbind"


Differential Revision: [D20525680](https://our.internmc.facebook.com/intern/diff/D20525680)

[ghstack-poisoned]

Commit:
Update on "Doxygen for torchbind"


![image](https://user-images.githubusercontent.com/4685384/77018550-1e444900-693b-11ea-8dfa-828e764130b2.png)
![image](https://user-images.githubusercontent.com/4685384/77018556-23a19380-693b-11ea-9e71-5c07edd19ff5.png)
![image](https://user-images.githubusercontent.com/4685384/77018564-2bf9ce80-693b-11ea-897a-16972687f2a4.png)
![image](https://user-images.githubusercontent.com/4685384/77018573-31571900-693b-11ea-9852-065f7f99fd2c.png)


Differential Revision: [D20525680](https://our.internmc.facebook.com/intern/diff/D20525680)

[ghstack-poisoned]

dzhulgakov commented on torch/custom_class.h:
```diff
@@ -17,52 +17,44 @@
+  /// see this class exposed as in Python and TorchScript. For example, if
+  /// you pass in "MyStack" here, the class will appear as
+  /// `torch.classes.MyStack` in both Python and TorchScript.
   class_(std::string className_) : className(std::move(className_)) {
```
stupid question - why is it className_ (with underscore)? :)

should the constructor be `explicit`?

jamesr66a replied:
1) Dunno, horace wrote it like that. I can change it
2) yeah probably

dzhulgakov commented on torch/custom_class.h:
```diff
@@ -96,46 +88,47 @@ class class_ {
+  /// where `CurClass` is the class you're registering and `T` is some object
+  /// that encapsulates the state of the object.
+  ///
+  ///     __getstate__(intrusive_ptr<CurClass>) -> T
```
maybe add example of actual registration (i.e. def_pickle call)?

jlin27 reviewed (approved):
The doxygen build changes LGTM

Commit:
Update on "Doxygen for torchbind"


![image](https://user-images.githubusercontent.com/4685384/77018550-1e444900-693b-11ea-8dfa-828e764130b2.png)
![image](https://user-images.githubusercontent.com/4685384/77018556-23a19380-693b-11ea-9e71-5c07edd19ff5.png)
![image](https://user-images.githubusercontent.com/4685384/77018564-2bf9ce80-693b-11ea-897a-16972687f2a4.png)
![image](https://user-images.githubusercontent.com/4685384/77018573-31571900-693b-11ea-9852-065f7f99fd2c.png)


Differential Revision: [D20525680](https://our.internmc.facebook.com/intern/diff/D20525680)

[ghstack-poisoned]

Commit:
Update on "Doxygen for torchbind"


![image](https://user-images.githubusercontent.com/4685384/77018550-1e444900-693b-11ea-8dfa-828e764130b2.png)
![image](https://user-images.githubusercontent.com/4685384/77018556-23a19380-693b-11ea-9e71-5c07edd19ff5.png)
![image](https://user-images.githubusercontent.com/4685384/77018564-2bf9ce80-693b-11ea-897a-16972687f2a4.png)
![image](https://user-images.githubusercontent.com/4685384/77018573-31571900-693b-11ea-9852-065f7f99fd2c.png)


Differential Revision: [D20525680](https://our.internmc.facebook.com/intern/diff/D20525680)

[ghstack-poisoned]
